services/QuillLMS/app/models/student_response.rb

```diff
@@ -0,0 +1,257 @@
+    metadata = data_hash[:metadata]
+
+    # There are six different strings that we normalize, and we use the
+    # same process for all of them:
+    #  1) If the key for the string isn't provided, skip it
+    #  2) Find the ID for this exact string if we've seen it before
+    #  3) Create a new normalized record if this string is brand new
+    #  4) Link the normalized reference to the StudentResponse model
+    # This array of arrays is used to execute that logic on each of the
+    # six strings we care about without actually writing the logic six
+    # different times.
+
+    attribute_model_value_tuples = [
+      [:student_response_answer_text=,
+       StudentResponseAnswerText,
+       metadata[:answer]],
+      [:student_response_directions_text=,
+       StudentResponseDirectionsText,
+       metadata[:directions]],
+      [:student_response_instructions_text=,
+       StudentResponseInstructionsText,
+       metadata[:instructions]],
+      [:student_response_previous_feedback_text=,
+       StudentResponsePreviousFeedbackText,
+       metadata[:lastFeedback]],
+      [:student_response_prompt_text=,
+       StudentResponsePromptText,
+       metadata[:prompt]],
+      [:student_response_question_type=,
+       StudentResponseQuestionType,
+       data_hash[:question_type]]
+    ]
+
+    attribute_model_value_tuples.each do |attribute, model, value|
+      next if value.nil?
+
+      related_model = model.find_or_create_by(text: value)
+      target.send(attribute, related_model)
+    end
+  end
+
+  # Takes an input of hashes representing old-style ConceptResult
+  # and rolls them up so that a single response has only one entry
+  # plus an array of concept_ids that match to it
+  private_class_method def self.roll_up_concepts(data_hash_array)
+    data_hash_array.each_with_object({}) do |value, rollup|
+      key = "#{value[:metadata][:questionNumber]}-#{value[:metadata][:attemptNumber]}"
+      if rollup[key]
+        rollup[key][:concept_ids].push(value[:concept_id])
+      else
+        rollup[key] = value.merge(concept_ids: [value[:concept_id]])
+      end
+      rollup
+    end.values
+  end
+
+  private_class_method def self.calculate_question_from_hash(data_hash)
+    question_uid = data_hash[:question_uid] || data_hash[:questionUid]
+    if !question_uid && data_hash[:metadata][:questionNumber]
+      activity_session = ActivitySession.find(data_hash[:activity_session_id])
+      question_index = data_hash[:metadata][:questionNumber] - 1
+      question_uid = activity_session.activity.data['questions'][question_index]['key']
+    end
+
+    return if !question_uid
+
+    Question.find_by_uid(question_uid)
+  end
+
+  private def legacy_format_metadata
+    metadata = legacy_format_base_metadata
+    metadata.merge!(student_response_extra_metadata || {})
+    metadata.reject { |_,v| v.blank? }
+  end
+
+  private def legacy_format_base_metadata
+    metadata = {
```

Comment:
This `metadata=` is not used.
```suggestion
     {
```

Reply:
Good catch.  Removed.